lib/trie/database.go

```diff
@@ -61,241 +61,267 @@ func (t *Trie) store(db chaindb.Batch, curr Node) error {
 	}
 
 	k := codec.KeyLEToNibbles(key)
 
-	enc, err := db.Get(root[:])
+	encodedRootNode, err := db.Get(rootHash[:])
 	if err != nil {
-		return nil, fmt.Errorf("failed to find root key=%s: %w", root, err)
+		return nil, fmt.Errorf("cannot find root hash key 0x%x: %w", rootHash, err)
 	}
 
-	rootNode, err := node.Decode(bytes.NewReader(enc))
+	reader := bytes.NewReader(encodedRootNode)
+	rootNode, err := node.Decode(reader)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot decode root node: %w", err)
 	}
 
 	return getFromDB(db, rootNode, k)
 }
 
-func getFromDB(db chaindb.Database, parent Node, key []byte) ([]byte, error) {
-	var value []byte
-
-	switch p := parent.(type) {
-	case *node.Branch:
-		length := lenCommonPrefix(p.Key, key)
-
-		// found the value at this node
-		if bytes.Equal(p.Key, key) || len(key) == 0 {
-			return p.Value, nil
-		}
-
-		// did not find value
-		if bytes.Equal(p.Key[:length], key) && len(key) < len(p.Key) {
-			return nil, nil
-		}
-
-		if p.Children[key[length]] == nil {
-			return nil, nil
+// getFromDB recursively searches through the trie and database
+// for the value corresponding to a key.
+// Note it does not copy the value so modifying the value bytes
+// slice will modify the value of the node in the trie.
+func getFromDB(db chaindb.Database, n Node, key []byte) (
+	value []byte, err error) {
+	leaf, ok := n.(*node.Leaf)
+	if ok {
+		if bytes.Equal(leaf.Key, key) {
+			return leaf.Value, nil
 		}
+		return nil, nil
+	}
 
-		// load child with potential value
-		enc, err := db.Get(p.Children[key[length]].GetHash())
-		if err != nil {
-			return nil, fmt.Errorf("failed to find node in database: %w", err)
-		}
+	branch := n.(*node.Branch)
+	// Key is equal to the key of this branch or is empty
+	if len(key) == 0 || bytes.Equal(branch.Key, key) {
+		return branch.Value, nil
+	}
 
-		child, err := node.Decode(bytes.NewReader(enc))
-		if err != nil {
-			return nil, err
-		}
+	commonPrefixLength := lenCommonPrefix(branch.Key, key)
+	if len(key) < len(branch.Key) && bytes.Equal(branch.Key[:commonPrefixLength], key) {
+		// The key to search is a prefix of the node key and is smaller than the node key.
+		// Example: key to search: 0xabcd
+		//          branch key:    0xabcdef
+		return nil, nil
+	}
 
-		value, err = getFromDB(db, child, key[length+1:])
-		if err != nil {
-			return nil, err
-		}
-	case *node.Leaf:
-		if bytes.Equal(p.Key, key) {
-			return p.Value, nil
-		}
-	case nil:
+	// childIndex is the nibble after the common prefix length in the key being searched.
+	childIndex := key[commonPrefixLength]
+	childWithHashOnly := branch.Children[childIndex]
+	if childWithHashOnly == nil {
 		return nil, nil
+	}
 
+	childHash := childWithHashOnly.GetHash()
+	encodedChild, err := db.Get(childHash)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"cannot find child with hash 0x%x in database: %w",
+			childHash, err)
 	}
-	return value, nil
+
+	reader := bytes.NewReader(encodedChild)
```

Comment:
Same as above, this is for readability. I used to nest calls to save on the number of lines but I now prefer to split in on more lines since I find it way more nice to read.